src/solvers/lowering/byte_operators.cpp

```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
+                                    : ID_byte_extract_big_endian;
+
+  const exprt subtype_size = size_of_expr(subtype, ns);
+
+  // compute the index of the first element of the array/vector that may be
+  // updated
+  exprt first_index = div_exprt(
+    typecast_exprt::conditional_cast(src.offset(), subtype_size.type()),
+    subtype_size);
+  simplify(first_index, ns);
+
+  // compute the offset within that first element
+  const exprt update_offset = mod_exprt(
+    typecast_exprt::conditional_cast(src.offset(), subtype_size.type()),
+    subtype_size);
+
+  // compute the number of bytes (from the update value) are going to be
+  // consumed for updating the first element
+  exprt initial_bytes = minus_exprt(subtype_size, update_offset);
+  exprt update_bound;
+  if(non_const_update_bound.has_value())
+    update_bound = typecast_exprt::conditional_cast(
+      *non_const_update_bound, subtype_size.type());
+  else
+    update_bound =
+      from_integer(value_as_byte_array.operands().size(), initial_bytes.type());
+  initial_bytes = if_exprt(
+    binary_predicate_exprt(initial_bytes, ID_lt, update_bound),
+    initial_bytes,
+    update_bound);
+  simplify(initial_bytes, ns);
+
+  // encode the first update: update the original element at first_index with
+  // initial_bytes-many bytes extracted from value_as_byte_array
+  with_exprt result(
+    src.op(),
+    first_index,
+    lower_byte_operators(
+      byte_update_exprt(
+        src.id(),
+        index_exprt(src.op(), first_index),
+        update_offset,
+        byte_extract_exprt(
+          extract_opcode,
+          value_as_byte_array,
+          from_integer(0, src.offset().type()),
+          array_typet(unsignedbv_typet(8), initial_bytes))),
+      ns));
+
+  // we will update one array/vector element at a time - compute the number of
+  // update values that will be consumed in each step; if we cannot determine a
+  // constant value at this time we assume it's just one byte
+  std::size_t step_size = 1;
+  if(subtype_size.is_constant())
+    step_size = numeric_cast_v<std::size_t>(to_constant_expr(subtype_size));
+  // given the first update already encoded, keep track of the offset into the
+  // update value
+  std::size_t offset = 0;
+  if(initial_bytes.is_constant())
+    offset = numeric_cast_v<std::size_t>(to_constant_expr(initial_bytes));
+
+  std::size_t i = 1;
+  for(; offset + step_size <= value_as_byte_array.operands().size();
+      offset += step_size)
+  {
+    exprt where = simplify_expr(
+      plus_exprt(first_index, from_integer(i, first_index.type())), ns);
+    ++i;
+
+    exprt element = lower_byte_operators(
+      byte_update_exprt(
+        src.id(),
+        index_exprt(src.op(), where),
+        from_integer(0, src.offset().type()),
+        byte_extract_exprt(
+          extract_opcode,
+          value_as_byte_array,
+          from_integer(offset, src.offset().type()),
+          array_typet(unsignedbv_typet(8), subtype_size))),
```

Comment:
It's a little surprising to see `byte_extract` used to pick up a slice of a byte-array; is `extractbits_exprt` better for this, or just an `array_exprt` of the relevant bytes? It's not better or worse particularly, I just note we don't need byte-extract's casting powers here.

Reply:
I need/want the power of a possibly non-constant size of the slice, hence falling back to `byte_extract` (which in turn is subjected to lowering, which might result in either an `extractbits_exprt` or `array_exprt`).